### ISSUE TITLE
`--target` not always applied

### ISSUE THREAD

**Issue by Finomnis** [foresterre/cargo-msrv#936]
## TLDR
`cargo msrv --output-format json --all-features verify --target thumbv7em-none-eabihf` does not respect the `--target` argument if it is behind the `verify` command.

## Details

In my case, I would like to combine `cargo-minimal-versions` with `cargo-msrv`. I use this command line:

```bash
cargo minimal-versions msrv --output-format json --all-features --target thumbv7em-none-eabihf verify
```

This internally runs:
- `cargo hack msrv --output-format json --all-features --target thumbv7em-none-eabihf verify`
- `cargo msrv --output-format json --all-features verify --target thumbv7em-none-eabihf`

Which then causes:
```json
{"type":"meta","instance":"cargo-msrv","version":"0.16.0-beta.22","sha_short":null,"target_triple":"x86_64-pc-windows-msvc","cargo_features":"default,rust_releases_dist_source","rustc":"1.79.0"}
```
... dropping the `thumbv7em-none-eabihf`, albeit not causing an error/warning.

I don't know why `cargo hack` switches the order of the arguments, but it seems that `cargo-msrv` does not respect the `--target` argument if it comes **after** the `verify` command.

**Comment by foresterre** [comment]
The target argument is possibly defined on the msrv command, and not on verify, and is not marked as global. Altough it would be odd that it is dropped silently...

**Comment by foresterre** [comment]
> The target argument is possibly defined on the msrv command, and not on verify, and is not marked as global. Although it would be odd that it is dropped silently...

So I checked it and `--target` is defined on `cargo msrv` level indeed (and not marked as global). As such it is currently only available before the subcommand.

I've to dive a bit deeper into why `--target` after the verify is silently ignored, and if it would be logical to (and how to) extend `cargo msrv verify` support to include `--target` after the subcommand.

**Comment by foresterre** [comment]
> > The target argument is possibly defined on the msrv command, and not on verify, and is not marked as global. Although it would be odd that it is dropped silently...
> 
> So I checked it and `--target` is defined on `cargo msrv` level indeed (and not marked as global). As such it is currently only available before the subcommand.
> 
> I've to dive a bit deeper into why `--target` after the verify is silently ignored, and if it would be logical to (and how to) extend `cargo msrv verify` support to include `--target` after the subcommand.

- If `cargo msrv --target x verify --target y` is given then `target = x` is used.
- If `cargo msrv --target x verify` is given then `target = x` is used.
- If `cargo msrv verify --target y` is given then `target = (default)` is used.
  - I'm not sure yet why `--target` after `verify` is not rejected
  - This seems to be equivalent to `cargo msrv verify`

**Comment by Finomnis** [comment]
> and if it would be logical to (and how to) extend `cargo msrv verify` support to include `--target` after the subcommand.

If that's possible, I would welcome that :)

**Comment by foresterre** [comment]
Found the culprit. The primary distraction comes from the issue that the primary command "cargo msrv" is defined at the top level, while all other commands are defined as subcommands; this makes sharing cli arguments harder (I intend to change this in the future, but for some backwards compatibility reasons it currently is what it is).

As such, these options: 

- RustReleasesOpts:
  - Find (top level): https://github.com/foresterre/cargo-msrv/blob/main/src/cli/find_opts.rs#L56
  - Verify: https://github.com/foresterre/cargo-msrv/blob/main/src/cli.rs#L147
- ToolchainOpts:
  - Find (top level): https://github.com/foresterre/cargo-msrv/blob/main/src/cli/find_opts.rs#L59
  - Verify: https://github.com/foresterre/cargo-msrv/blob/main/src/cli.rs#L150 
- CheckCommandOpts:
  - Find (top level): https://github.com/foresterre/cargo-msrv/blob/main/src/cli/find_opts.rs#L62
  - Verify: https://github.com/foresterre/cargo-msrv/blob/main/src/cli.rs#L153

Exist on both the top level "find msrv" command, and the subcommand `verify`.
The implementation however only reads from the top level. The subcommand variants are ignored.

By having these defined on the `verify` subcommand, you could provide values for these arguments, without the program rejecting these. However, since we don't read from these, you experience this "silently ignore" behaviour.

**Comment by foresterre** [comment]
I wrote a few too many words about it this morning: https://gribnau.dev/posts/puzzle-sharing-declarative-args-between-top-level-and-subcommand/

I'm considering how to best solve this. I might just patch only the `target` field for now until I figure out a better solution.

**Comment by Finomnis** [comment]
:D blogging addicted, eh?

Sure, thanks so much for your efforts :)

**Comment by foresterre** [comment]
I've released it in https://github.com/foresterre/cargo-msrv/releases/tag/v0.16.0-beta.23.

Relevant known issues are documented in the PR https://github.com/foresterre/cargo-msrv/pull/943 and the https://github.com/foresterre/cargo-msrv/blob/main/CHANGELOG.md#known-issues

**Comment by Finomnis** [comment]
Great! Thank you very much!
